File: src/carve/_grids.py
"""Default estimator and preprocessing grids for CARVE."""

from typing import Literal

import numpy as np
from sklearn.cluster import HDBSCAN, KMeans, AgglomerativeClustering
from sklearn.decomposition import PCA
from sklearn.manifold import TSNE
from sklearn.neighbors import NearestNeighbors
from sklearn.preprocessing import FunctionTransformer, StandardScaler
from umap import UMAP

from .cluster import LeidenClustering, LouvainClustering, SpectralClustering
from ._sweep import SweepSpec, resolve_sweep
from ._types import GridSpec, PreprocSpec


def estimate_knn_gamma(
    X: np.ndarray,
    n_neighbors: int = 7,
    multipliers: tuple[float, ...] = (0.5, 1.0, 2.0),
) -> list[float]:
    """Estimate RBF gamma values using a k-NN median heuristic.

    Fits a NearestNeighbors model, takes the k-th neighbor distance for
    each point, computes sigma = median(kth_distances), and returns
    gamma = 1 / (2 * (m * sigma)^2) for each multiplier m.

    Parameters
    ----------
    X : ndarray of shape (n_samples, n_features)
        Input data.
    n_neighbors : int, default=7
        Number of neighbors.
    multipliers : tuple of float, default=(0.5, 1.0, 2.0)
        Scale multipliers applied to sigma.

    Returns
    -------
    gammas : list of float
        Gamma values spanning a range around the local scale.
    """
    nn = NearestNeighbors(n_neighbors=n_neighbors)
    nn.fit(X)
    dists, _ = nn.kneighbors(X)
    kth_dists = dists[:, -1]
    sigma = float(np.median(kth_dists))
    if sigma <= 0:
        sigma = (
            float(np.mean(kth_dists[kth_dists > 0])) if np.any(kth_dists > 0) else 1.0
        )
    return [float(1.0 / (2.0 * (m * sigma) ** 2)) for m in multipliers]


def default_estimator_grids(
    X: np.ndarray,
    n_clusters: int | np.ndarray = 10,
    preset: Literal["light", "full"] = "light",
    *,
    sweep: SweepSpec | None = None,
) -> list[GridSpec]:
    """Return default clustering estimator grids for a run's sweep axis.

    Parameters
    ----------
    X : ndarray of shape (n_samples, n_features)
        Input data used to derive data-driven hyperparameters (e.g., gamma).
    n_clusters : int or ndarray, default=10
        Number(s) of clusters to evaluate. Ignored when ``sweep`` is given.
    preset : {"light", "full"}, default="light"
        ``"light"`` includes KMeans, Ward-linkage agglomerative, and
        self-tuning spectral clustering. ``"full"`` adds average- and
        single-linkage agglomerative clustering and RBF-kernel spectral
        clustering with data-driven gamma.
    sweep : SweepSpec, optional
        Resolved sweep axis. Defaults to an ``n_clusters`` sweep built
        from n_clusters.

    Returns
    -------
    grids : list of tuple
        List of (EstimatorClass, param_grid) tuples suitable for
        ``sklearn.model_selection.ParameterGrid``.

    Raises
    ------
    ValueError
        If no default grids exist for the sweep parameter.
    """
    if sweep is None:
        sweep = resolve_sweep(n_clusters=n_clusters)

    if sweep.param == "n_clusters":
        return _default_k_grids(X, sweep.values, preset)
    if sweep.param == "resolution":
        return _default_resolution_grids(sweep.values, preset)
    if sweep.param == "min_cluster_size":
        return _default_min_cluster_size_grids(sweep.values, preset)

    raise ValueError(
        f"No default estimator grids for sweep parameter {sweep.param!r}. "
        "Pass estimator_param_grids=... explicitly."
    )


def _default_k_grids(
    X: np.ndarray,
    n_clusters: np.ndarray,
    preset: Literal["light", "full"] = "light",
) -> list[GridSpec]:
    """Grids for estimators that take an explicit number of clusters."""
    ks = list(np.asarray(n_clusters).tolist())

    grids: list[GridSpec] = [
        (KMeans, {"n_clusters": ks}),
        (AgglomerativeClustering, {"n_clusters": ks, "linkage": ["ward"]}),
        (SpectralClustering, {"n_clusters": ks, "affinity": ["self_tuning"]}),
    ]

    if preset == "full":
        grids.append(
            (
                AgglomerativeClustering,
                {"n_clusters": ks, "linkage": ["average", "single", "complete"]},
            ),
        )

        grids.append(
            (
                SpectralClustering,
                {"n_clusters": ks, "affinity": ["rbf"], "gamma": estimate_knn_gamma(X)},
            ),
        )

    return grids


def _default_resolution_grids(
    resolutions: np.ndarray,
    preset: Literal["light", "full"] = "light",
) -> list[GridSpec]:
    """Grids for graph-community estimators swept over ``resolution``.

    Parameters
    ----------
    resolutions : ndarray
        Resolution values to evaluate.
    preset : {"light", "full"}, default="light"
        ``"light"`` uses Leiden and Louvain on a 15-neighbor graph.
        ``"full"`` additionally varies the neighborhood size.

    Returns
    -------
    grids : list of tuple
        List of (EstimatorClass, param_grid) tuples.
    """
    res = [float(r) for r in np.asarray(resolutions).tolist()]

    grids: list[GridSpec] = [
        (LeidenClustering, {"resolution": res, "n_neighbors": [15]}),
        (LouvainClustering, {"resolution": res, "n_neighbors": [15]}),
    ]

    if preset == "full":
        grids.append((LeidenClustering, {"resolution": res, "n_neighbors": [10, 30]}))
        grids.append((LouvainClustering, {"resolution": res, "n_neighbors": [10, 30]}))

    return grids


def _default_min_cluster_size_grids(
    min_cluster_sizes: np.ndarray,
    preset: Literal["light", "full"] = "light",
) -> list[GridSpec]:
    """Grids for HDBSCAN swept over ``min_cluster_size``.

    Parameters
    ----------
    min_cluster_sizes : ndarray
        Minimum cluster sizes to evaluate.
    preset : {"light", "full"}, default="light"
        ``"full"`` additionally evaluates leaf cluster selection.

    Returns
    -------
    grids : list of tuple
        List of (EstimatorClass, param_grid) tuples.
    """
    sizes = [int(s) for s in np.asarray(min_cluster_sizes).tolist()]

    grids: list[GridSpec] = [
        (
            HDBSCAN,
            {"min_cluster_size": sizes, "cluster_selection_method": ["eom"]},
        ),
    ]

    if preset == "full":
        grids.append(
            (
                HDBSCAN,
                {"min_cluster_size": sizes, "cluster_selection_method": ["leaf"]},
            )
        )

    return grids


def default_normalization_options() -> list[PreprocSpec]:
    """Return default normalization preprocessing options.

    Returns
    -------
    options : list of tuple
        List of (TransformerClass, param_grid) pairs.
    """
    return [
        (FunctionTransformer, {}),
        (StandardScaler, {}),
        (FunctionTransformer, {"func": [np.log1p]}),
    ]


def default_dim_reduction_options(
    X: np.ndarray,
    subsample_ratio: float = 0.6,
) -> list[PreprocSpec]:
    """Return default dimensionality reduction options.

    Parameters
    ----------
    X : ndarray of shape (n_samples, n_features)
        Input data used to determine feasible component counts.
    subsample_ratio : float, default=0.6
        Subsample ratio used in validation; constrains DR hyperparameters.

    Returns
    -------
    options : list of tuple
        List of (TransformerClass, param_grid) pairs.
    """
    n_samples, p = X.shape
    min_n = int(round(n_samples * (1 - subsample_ratio))) - 1

    return [
        (FunctionTransformer, {}),
        (PCA, {"n_components": list(range(2, min(min_n, p)))}),
        (TSNE, {"n_components": [2], "perplexity": list(range(5, min(min_n, 51)))}),
        (
            UMAP,
            {
                "n_components": list(range(2, min(min_n, p))),
                "n_neighbors": list(range(5, 51)),
                "min_dist": [0.1],
            },
        ),
    ]

File: carve-r/R/grids.R
# Default estimator and preprocessing grids. Mirrors _grids.py.
#
# Grid-spec form in R:
#
#   list(type = "<backend>", grid = list(param = values, ...))
#
# `type` is forwarded to `cluster_labels`. `grid` is a named list of
# parameter *values*, each either a scalar or a vector; `expand.grid`
# materializes one config per combination.
#
# Preprocessor-spec form (normalization / dimensionality reduction):
#
#   list(type = "<kind>", grid = list(param = values), name = "<display>")
#
# Currently supported types: "identity", "standardize", "log1p", "pca".


#' Estimate RBF gamma values from a k-NN median heuristic
#'
#' Mirrors `_grids.estimate_knn_gamma`. Computes the median k-th neighbor
#' distance and returns `gamma = 1 / (2 * (m * sigma)^2)` for each `m` in
#' `multipliers`.
#'
#' @param X Numeric matrix of shape `(n_samples, n_features)`.
#' @param n_neighbors Integer; number of neighbors (default `7`).
#' @param multipliers Numeric vector of scale multipliers applied to sigma.
#'
#' @return Numeric vector of gamma values.
#'
#' @export
estimate_knn_gamma <- function(X, n_neighbors = 7L,
                               multipliers = c(0.5, 1.0, 2.0)) {
  X <- ensure_2d_matrix(X)
  k_excl <- max(1L, as.integer(n_neighbors) - 1L)
  knn <- FNN::get.knn(X, k = k_excl, algorithm = "kd_tree")
  kth <- knn$nn.dist[, k_excl]

  sigma <- stats::median(kth)
  if (sigma <= 0) {
    pos <- kth[kth > 0]
    sigma <- if (length(pos) > 0L) mean(pos) else 1
  }
  as.numeric(1 / (2 * (multipliers * sigma)^2))
}


#' Default estimator grids
#'
#' Mirrors `_grids.default_estimator_grids`. Returns a list of grid specs
#' suitable for iteration in [run_validation()].
#'
#' @param X Numeric matrix used to derive data-driven hyperparameters
#'   (currently unused for `"light"`).
#' @param n_clusters Integer vector or scalar; candidate cluster counts.
#' @param preset `"light"` (KMeans + Ward agglomerative + self-tuning
#'   spectral) or `"full"` (adds average/single/complete linkage and
#'   RBF-kernel spectral with data-driven `gamma`).
#'
#' @return A list of grid specs.
#'
#' @export
default_estimator_grids <- function(X, n_clusters = 10L,
                                    preset = c("light", "full")) {
  preset <- match.arg(preset)
  ks <- coerce_n_clusters(n_clusters)

  grids <- list(
    list(type = "kmeans", grid = list(n_clusters = ks)),
    list(type = "agglomerative",
         grid = list(n_clusters = ks, linkage = "ward")),
    list(type = "spectral",
         grid = list(n_clusters = ks, affinity = "self_tuning"))
  )

  if (identical(preset, "full")) {
    grids <- c(
      grids,
      list(
        list(type = "agglomerative",
             grid = list(n_clusters = ks,
                         linkage = c("average", "single", "complete"))),
        list(type = "spectral",
             grid = list(n_clusters = ks, affinity = "rbf",
                         gamma = estimate_knn_gamma(X)))
      )
    )
  }

  grids
}


#' Default normalization options
#'
#' Mirrors `_grids.default_normalization_options`. Returns identity,
#' standardization, and `log1p`.
#'
#' @return A list of preprocessor specs.
#'
#' @export
default_normalization_options <- function() {
  list(
    list(type = "identity", grid = list(), name = "Identity"),
    list(type = "standardize", grid = list(), name = "StandardScaler"),
    list(type = "log1p", grid = list(), name = "Log1p")
  )
}


#' Default dimensionality reduction options
#'
#' Mirrors `_grids.default_dim_reduction_options` but only supports
#' identity and PCA in this phase; t-SNE and UMAP are reserved for a
#' later milestone. The PCA grid spans `2:min(min_n, p)` components,
#' where `min_n = round(n * (1 - subsample_ratio)) - 1` is the smallest
#' subsample size CARVE will use downstream.
#'
#' @param X Numeric matrix.
#' @param subsample_ratio Numeric in `(0, 1)`.
#'
#' @return A list of preprocessor specs.
#'
#' @export
default_dim_reduction_options <- function(X, subsample_ratio = 0.6) {
  X <- ensure_2d_matrix(X)
  n <- nrow(X)
  p <- ncol(X)
  min_n <- as.integer(round(n * (1 - subsample_ratio))) - 1L
  upper <- min(min_n, p)
  components <- if (upper >= 2L) seq.int(2L, upper) else integer(0L)

  opts <- list(
    list(type = "identity", grid = list(), name = "Identity")
  )
  if (length(components) > 0L) {
    opts <- c(opts, list(
      list(type = "pca",
           grid = list(n_components = components),
           name = "PCA")
    ))
  }
  opts
}


# Expand a grid-spec `grid` into a list of parameter lists. Mirrors
# sklearn's `ParameterGrid`. Returns list() for an empty grid (one empty
# config, matching ParameterGrid semantics).
expand_grid_spec <- function(grid) {
  if (length(grid) == 0L) return(list(list()))
  # Preserve argument order; stringsAsFactors=FALSE keeps character values.
  df <- do.call(
    expand.grid,
    c(grid, list(stringsAsFactors = FALSE, KEEP.OUT.ATTRS = FALSE))
  )
  lapply(seq_len(nrow(df)), function(i) as.list(df[i, , drop = FALSE]))
}
